imap/cyr_guesstzdb.c

```diff
@@ -0,0 +1,187 @@
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
+ * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
+ * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <config.h>
+
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+#include <stdlib.h>
+#include <stdio.h>
+#include <sysexits.h>
+
+#include "cyrusdb.h"
+#include "global.h"
+#include "proc.h"
+#include "../master/masterconf.h"
+
+#include "guesstz.h"
+
+/* Make ld happy */
+const char *MASTER_CONFIG_FILENAME = DEFAULT_MASTER_CONFIG_FILENAME;
+
+/* Print usage info on stderr and exit */
+__attribute__((noreturn)) static int usage(const char *name)
+{
+    fprintf(stderr,
+            "usage: %s [mode] [options]\n"
+            "\n"
+            "Mode flags: \n"
+            "  -c                create database\n"
+            "  -p                print database\n"
+            "  -h                show usage\n"
+            "\n"
+            "General options:\n"
+            "  -C <alt-config>   use alternative imapd.conf config file\n"
+            "  -F <alt-file>     use alternative database file\n"
+            "Create mode options:\n"
+            "  -R <start>,<end>  cover time range\n"
+            "  -Z <alt-zoneinfo> use alternative zoneinfo dir\n"
+            "\n",
+        name);
+    exit(EX_USAGE);
+}
+
+int main(int argc, char *argv[])
+{
+    char *alt_config = NULL;
+    const char *alt_fname = NULL;
+    const char *alt_range = NULL;
+    const char *alt_zoneinfo = NULL;
+    enum mode { CREATE, PRINT, UNKNOWN } m = UNKNOWN;
+    char *namec = strdup(argv[0]);
+    char *name = basename(namec);
+
+    extern char *optarg;
+    int opt;
+
+    /* Parse arguments */
+    while ((opt = getopt(argc, argv, "chpC:F:R:Z:")) != EOF) {
+        switch (opt) {
+        case 'C':               /* alt config file */
+            alt_config = optarg;
+            break;
+        case 'F':
+            alt_fname = optarg;
+            break;
+        case 'R':
+            alt_range = optarg;
+            break;
+        case 'Z':
+            alt_zoneinfo = optarg;
+            break;
+        case 'c':
+            if (m != UNKNOWN)
+                usage(name);
+            m = CREATE;
+            break;
+        case 'h':
+            usage(name);
+            break;
+        case 'p':
+            if (m != UNKNOWN)
+                usage(name);
+            m = PRINT;
+            break;
+        default:
+            usage(name);
+        }
+    }
+    if (optind < argc || m == UNKNOWN) {
+        usage(name);
+    }
+
+    cyrus_init(alt_config, "cyr_guesstzdb", 0, 0);
+
+    if (m == CREATE) {
+        /* Determine observance expansion time range */
+        const icaltimezone *utc = icaltimezone_get_utc_timezone();
+        icaltimetype dbstart = {
+            2000, 1, 1, 0, 0, 0, 0, 0, utc
+        };
+        icaltimetype dbend = {
+            2038, 1, 1, 0, 0, 0, 0, 0, utc
+        };
```

Comment:
Not sure how future-proof this type of initialization might be.  Might be easier to just use icaltime_from_string() with a const char *

Reply:
Done